design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,574 @@
+# KLIP 24 - KEY column semantics in queries
+
+**Author**: @big-andy-coates |
+**Release Target**: TBD |
+**Status**: In Discussion |
+**Discussion**: TBD
+
+**tl;dr:** Persistent queries do not allow the key column in the projection as key columns are
+currently _implicitly_ copied across. This is not intuitive to anyone familiar with SQL. This KLIP
+proposes flipping this so that the key column is _not_ implicitly copied across.
+
+## Motivation and background
+
+Below, we contrast the differences between transient and persistent queries and their current key
+column handling semantics.
+
+Note: in all the example sql in this KLIP:
+ * `EMIT CHANGES` had been removed for brevity.
+ * `ROWTIME` had been ignored for brevity.
+ * Schemas use `KEY` to denote the column stored in the Kafka record's key.
+ * `PRIMARY KEY` is not used for brevity.
+ * All input sources use the schema `ID INT KEY, V0 INT, V1 INT`, unless otherwise stated.
+
+Transient push and pull queries only return the columns in their projection. This is inline with
+standard SQL. Conversely, persistent queries _implicitly_ copy the key column across, and don't
+allow the key column in the projection.
+
+### Simple query
+
+The difference in key column handling can be seen with a simple query:
+
+```sql
+SELECT ID, V0, V1 FROM INPUT;
+-- resulting columns: ID INT, V0 INT, V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS
+   SELECT ID, V0, V1 FROM INPUT;
+-- fails with error about duplicate 'ID' column.
+```
+
+Where as the above transient query works and returns all columns, converting it to a persistent
+query causes it to fail with a duplicate column error, because `ID` is _implicitly_ being copied to
```

Comment:
typo: `it to fail`

Reply:
I don't see the typo... can you spell it out please?

Reply:
Not sure any more... Sorry.